csrc/dynamic_transform.cpp

```diff
@@ -154,6 +181,97 @@ class DynamicTransformInitialInfoBuilder : public IterVisitor {
   std::vector<Val*> leaf_dynamic_vals_;
 };
 
+//! This performs a depth-first search from outputs toward inputs for empty
+//! tensors. It does not traverse past any zero tensors it finds; this is why
+//! this is implemented as a single function instead of with BackwardVisitor.
```

Comment:
Originally I used `BackwardVisitor`, but when we encounter an empty tensor, we want to DCE its upstream branch at concretization. That means we should not continue processing it, because any dynamic transforms in the dead branches will lead to spurious concretization info that might cause cache misses due to different concretization of the dead branches. So I just process it with a stack instead so I can skip those branches easily.

Reply:
I'm fine with it as is, but sounds like you could do so by extending the `next` method. See https://github.com/NVIDIA/Fuser/blob/main/csrc/iter_visitor.cpp#L308-L313, for example.

Reply:
The problem is that `next` works in the forward direction: it tells you the next downstream statements to process. In `BackwardVisitor`, this is used in a forward traversal from `InputsOf(outputs)` to outputs to build a stack of statements, then the statements are handled by popping from the stack, which reverses the handling. https://github.com/NVIDIA/Fuser/blob/e2bc9f39fc22afeb76e7ba1f652b78728851c447/csrc/iter_visitor.cpp#L460-L480. The original `InputsOf` is determining the roots to flow forward from. Instead, I just build the stack during a single backward pass that finds terminates at empty tensors, which would otherwise require modifying the behavior of `Inputs` inside `BackwardVisitor`.